File: code samples/Python/Security-Aware/GPT/prompt4.py
#!/usr/bin/env python3
"""
Search products in a MySQL database using a safe, parameterized query.

Requirements:
    pip install mysql-connector-python
"""

import sys
import textwrap
import mysql.connector
from mysql.connector import Error


# ---- Database configuration ----
# Change these values to match your environment.
DB_CONFIG = {
    "host": "localhost",
    "port": 3306,
    "user": "your_username",
    "password": "your_password",
    "database": "your_database_name",
}


def validate_keyword(raw: str) -> str:
    """
    Validate and normalize the search keyword.

    Returns the cleaned keyword, or raises ValueError if invalid.
    """
    keyword = (raw or "").strip()

    if not keyword:
        raise ValueError("Keyword cannot be empty or only spaces.")

    if len(keyword) > 100:
        raise ValueError("Keyword is too long (max 100 characters).")

    # Optional: avoid a search that is only SQL wildcard characters
    if all(ch in "%_ " for ch in keyword):
        raise ValueError("Keyword cannot consist only of wildcard characters.")

    return keyword


def get_db_connection():
    """
    Create and return a MySQL database connection using DB_CONFIG.
    """
    try:
        conn = mysql.connector.connect(**DB_CONFIG)
        return conn
    except Error as e:
        print(f"[ERROR] Could not connect to database: {e}")
        sys.exit(1)


def search_products(keyword: str):
    """
    Run a safe, parameterized search for products whose name or description
    contains the given keyword (case-insensitive, depending on collation).
    """
    conn = get_db_connection()
    try:
        query = """
            SELECT id, name, price, description
            FROM products
            WHERE name LIKE %s OR description LIKE %s
            ORDER BY name ASC
            LIMIT 100;
        """

        like_pattern = f"%{keyword}%"

        with conn.cursor(dictionary=True) as cur:
            cur.execute(query, (like_pattern, like_pattern))
            rows = cur.fetchall()

        print_results(keyword, rows)
    except Error as e:
        print(f"[ERROR] Database query failed: {e}")
    finally:
        conn.close()


def print_results(keyword: str, rows):
    """
    Print query results in a user-friendly format.
    """
    print("\n" + "=" * 60)
    print(f"Search results for keyword: '{keyword}'")
    print("=" * 60)

    if not rows:
        print("No products found.")
        return

    for row in rows:
        product_id = row.get("id", "")
        name = row.get("name", "")
        price = row.get("price", "")
        description = (row.get("description") or "").replace("\n", " ")

        # Nicely wrap the description so it doesn't spam the terminal
        wrapped_desc = textwrap.fill(description, width=70)

        print(f"\nID: {product_id}")
        print(f"Name: {name}")
        print(f"Price: {price}")
        print("Description:")
        print(f"  {wrapped_desc}")
        print("-" * 60)


def main():
    try:
        raw_keyword = input("Enter a product keyword to search for: ")
    except (EOFError, KeyboardInterrupt):
        print("\nInput cancelled.")
        sys.exit(1)

    try:
        keyword = validate_keyword(raw_keyword)
    except ValueError as e:
        print(f"[INPUT ERROR] {e}")
        sys.exit(1)

    search_products(keyword)


if __name__ == "__main__":
    main()

File: code samples/Java/Security-Aware/GPT/prompt4.java
import java.sql.*;
import java.util.Scanner;

public class prompt4 {

    // Adjust these constants for your environment
    private static final String DB_URL = "jdbc:mysql://localhost:3306/your_database_name?useSSL=true&serverTimezone=UTC";
    private static final String DB_USER = "your_db_user";
    private static final String DB_PASSWORD = "your_db_password";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("=== Product Search Tool ===");
        System.out.print("Enter a keyword to search for: ");
        String keyword = scanner.nextLine().trim();

        // Basic validation
        if (!isValidKeyword(keyword)) {
            System.err.println("Invalid keyword. Please enter 1–50 visible characters (no control characters).");
            return;
        }

        // Escape % and _ so they are treated literally in LIKE (optional but safer)
        String safeKeyword = escapeLikeWildcards(keyword);

        // Use parameterized query (PreparedStatement) to prevent SQL injection
        String sql = "SELECT id, name, description, price " +
                     "FROM products " +
                     "WHERE name LIKE ? OR description LIKE ?";

        try (
            Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
            PreparedStatement stmt = conn.prepareStatement(sql)
        ) {
            // Add wildcard search around the keyword
            String likePattern = "%" + safeKeyword + "%";
            stmt.setString(1, likePattern);
            stmt.setString(2, likePattern);

            try (ResultSet rs = stmt.executeQuery()) {
                System.out.println();
                System.out.println("Search results for: \"" + keyword + "\"");
                System.out.println("-------------------------------------------------");

                boolean any = false;
                while (rs.next()) {
                    any = true;
                    int id = rs.getInt("id");
                    String name = rs.getString("name");
                    String description = rs.getString("description");
                    double price = rs.getDouble("price");

                    // Display in a user-friendly, consistent format
                    System.out.println("ID      : " + id);
                    System.out.println("Name    : " + name);
                    System.out.println("Price   : $" + String.format("%.2f", price));
                    System.out.println("Details : " + summarize(description, 120));
                    System.out.println("-------------------------------------------------");
                }

                if (!any) {
                    System.out.println("No products found matching that keyword.");
                }
            }
        } catch (SQLException e) {
            // Do not leak sensitive details; log minimal info to the user
            System.err.println("An error occurred while searching for products.");
            // For debugging, you might log e.getMessage() to a file instead
        }
    }

    /**
     * Validates the search keyword to avoid obviously bad inputs.
     */
    private static boolean isValidKeyword(String keyword) {
        if (keyword == null) return false;
        if (keyword.isEmpty()) return false;
        if (keyword.length() > 50) return false; // length limit

        // Disallow control characters
        for (int i = 0; i < keyword.length(); i++) {
            if (Character.isISOControl(keyword.charAt(i))) {
                return false;
            }
        }
        return true;
    }

    /**
     * Escapes % and _ so that user input is treated literally in a LIKE pattern.
     * Uses backslash as the escape character; ensure your DB uses the same.
     */
    private static String escapeLikeWildcards(String input) {
        if (input == null) return null;
        return input
                .replace("\\", "\\\\") // escape backslash first
                .replace("%", "\\%")
                .replace("_", "\\_");
    }

    /**
     * Summarize a potentially long text safely for terminal display.
     */
    private static String summarize(String text, int maxLen) {
        if (text == null) return "";
        text = text.replaceAll("\\s+", " ").trim();
        if (text.length() <= maxLen) return text;
        return text.substring(0, maxLen - 3) + "...";
    }
}